packages/sanes-chrome-extension/src/extension/background/model/signingServer/index.ts

```diff
@@ -0,0 +1,127 @@
+import { PublicIdentity, UnsignedTransaction } from '@iov/bcp';
+import { JsonRpcSigningServer, MultiChainSigner, SigningServerCore } from '@iov/core';
+import { JsonRpcResponse } from '@iov/jsonrpc';
+import { generateErrorResponse } from '../../errorResponseGenerator';
+import { getConfigurationFile } from '../persona/config';
+import { requestCallback } from './requestCallback';
+import {
+  GetIdentitiesRequest,
+  isRequestMeta,
+  Request,
+  RequestMeta,
+  RequestQueueManager,
+  SignAndPostRequest,
+} from './requestQueueManager';
+import { SenderWhitelist } from './senderWhitelist';
+
+export class SigningServer {
+  private requestHandler = new RequestQueueManager();
+  private senderWhitelist = new SenderWhitelist();
+  private signingServer: JsonRpcSigningServer | undefined;
+
+  public getIdentitiesCallback = (signer: MultiChainSigner) => async (
+    reason: string,
+    matchingIdentities: ReadonlyArray<PublicIdentity>,
+    meta: any, // eslint-disable-line
+  ) => {
+    if (!isRequestMeta(meta)) {
+      throw new Error('Unexpected type of data in meta field');
+    }
+    const { senderUrl } = meta;
+    const chainNames = (await getConfigurationFile()).names;
+
+    const requestedIdentities = matchingIdentities.map(matchedIdentity => {
+      const chainName = chainNames[matchedIdentity.chainId];
+
+      return {
+        name: chainName,
+        address: signer.identityToAddress(matchedIdentity),
+      };
+    });
+
+    const data: GetIdentitiesRequest = {
+      senderUrl,
+      requestedIdentities,
+    };
+
+    return requestCallback(
+      this.requestHandler,
+      this.senderWhitelist,
+      reason,
+      'getIdentities',
+      data,
+      matchingIdentities,
+      [],
+    );
+  };
+
+  public signAndPostCallback = () => (
+    reason: string,
+    transaction: UnsignedTransaction,
+    meta: any, // eslint-disable-line
+  ): Promise<boolean> => {
+    if (!isRequestMeta(meta)) {
+      throw new Error('Unexpected type of data in meta field');
+    }
+    const { senderUrl } = meta;
+
+    const data: SignAndPostRequest = {
+      senderUrl,
+      tx: transaction,
+    };
+
+    return requestCallback(
+      this.requestHandler,
+      this.senderWhitelist,
+      reason,
+      'signAndPost',
+      data,
+      true,
+      false,
+    );
+  };
+
+  public getPendingRequests(): Request[] {
+    return this.requestHandler.requests();
+  }
+
+  public start(core: SigningServerCore): void {
```

Comment:
This creates a circular dependency: Persona depends on SigningServer and SigningServer depends on Persona, which makes the flow very very hard to follow.

Reply:
I think the problem here is you are not familiar with it (yet). Just in the moment you know this SigningServer class is for dealing with external requests.

I know ideally should be independent, but at the end Persona and SigningServer should be connected somehow, because without Persona cannot exist SigningServer, and I prefer to just pass pack the "core" here rather than start the signing server in Persona which does not make really sense (inherited feature caused for bad design thinking persona should live in the app instead of the background).

Anyway this relation is just 1:1 from the main backgroundscript entity, so it does not hurt 😉